src/data/markdown/docs/40 xk6-disruptor/03 Examples/01 Inject HTTP faults into Pod.md

```diff
@@ -0,0 +1,435 @@
+
+```javascript
+ 47 export default function(data) {
+ 48     http.get(`http://${data.srv_ip}/delay/0.1`);
+ 49 }
+ ```
+
+<Blockquote mod="note">
+
+ The test uses the `delay` endpoint which return after the requested delay. It requests a `0.1s` (`100ms`) delay to ensure the baseline scenario (see scenarios below) has meaningful statistics for the request duration. If we were simply calling a locally deployed http server (for example `nginx`), the response time would exhibit a large variation between a few microseconds to a few milliseconds. Having `100ms` as baseline response time has proved to offer more consistent results.
+
+</Blockquote>
+
+## Fault injection
+
+The `disrupt` function creates a PodDisruptor](pod-disruptor) using a selector that matches pods in the namespace `httpbin-ns` with the label `app: httpbin` (lines 52-60). 
+
+The http faults are then injected by calling the `PodDisruptor`'s `injectHTTPFaults` method using a fault definition that introduces a delay of `50ms` on each request and an error code `500` in a `10%` of the requests (lines 63-68).
+
+```javascript
+ 51 export function disrupt(data) {
+ 52     const selector = {
+ 53         namespace: namespace,
+ 54             select: {
+ 55                 labels: {
+ 56                     app: app
+ 57                 }
+ 58         }
+ 59     }
+ 60     const podDisruptor = new PodDisruptor(selector)
+ 61
+ 62     // delay traffic from one random replica of the deployment
+ 63     const fault = {
+ 64         average_delay: 50,
+ 65         error_code: 500,
+ 66         error_rate: 0.1
+ 67     }
+ 68     podDisruptor.injectHTTPFaults(fault, 30)
+ 69 }
+```
+
+## Scenarios 
+
+This test defines three [scenarios](https://k6.io/docs/using-k6/scenarios) to be executed. The `base` scenario (lines 74-82) applies the test load to the target application for `30s` invoking the `default` function and it is used to set a baseline for the application's performance. The `disrupt` scenario (lines 83-89) is executed starting at the `30` seconds of the test run. It invokes once the `disrupt` function to inject a fault in the HTTP requests of the target application. The `faults` scenario (lines 90-98) is also executed starting at the `30` seconds of the test run, reproducing the same workload than the `base` scenario but now under the effect of the faults introduced by the `disrupt` scenario.
+
+```javascript
+ 73     scenarios: {
+ 74         base: {
+ 75             executor: 'constant-arrival-rate',
+ 76             rate: 100,
+ 77             preAllocatedVUs: 10,
+ 78             maxVUs: 100,
+ 79             exec: "default",
+ 80             startTime: '0s',
+ 81             duration: "30s",
+ 82         },
+ 83         disrupt: {
+ 84             executor: 'shared-iterations',
+ 85             iterations: 1,
+ 86             vus: 1,
+ 87             exec: "disrupt",
+ 88             startTime: "30s",
+ 89         },
+ 90         faults: {
```

Comment:
```
         'http_req_duration{scenario:base}': [],
         'http_req_duration{scenario:faults}': [],
         'http_req_failed{scenario:base}': [],
         'http_req_failed{scenario:faults}': [],
 ```

I continue reading and I interpret the thresholds are created like shown above to visualize the results on the terminal. 

I disagree with that approach. It seems the test implementation is driven by the shortcoming on the test visualization.

Reply:
> I continue reading and I interpret the thresholds are created like shown above to visualize the results on the terminal.
> I disagree with that approach. It seems the test implementation is driven by the shortcoming on the test visualization.

This is true but I don't see any alternative that does not make the example more complex. Do you have any suggestion?

Reply:
@pablochacin We had a similar discussion with the blog post.

Even though is not ideal, I would suggest:
- creating the test with the two scenarios
- if you want to mention thresholds, do it, otherwise skip Thresholds (I suggest skipping them)
- run the test twice: with and without running the disruptor scenario. You can mention it on the tutorial. 
- table comparing the metric results of both runs

Reply:
@ppcano I don't like the idea of having only one scenario for load and repeat the test. This is not the way I would like the disruptor to be used and I don't feel comfortable suggesting this in the example. The main reason is that unfortunately k6 does not offer a way for selecting/skipping scenarios and the way to implement this using environment variables is in my opinion too complex. 

As this is a topic that may require longer discussion I suggest to leave the example as it is and open an issue for discussing how to simplify it or make it more idiomatic.

Reply:
I have changed the example so it now has one load scenario and does not use the thresholds for reporting the metrics.